dataframe.py

```diff
@@ -0,0 +1,269 @@
+
+
+class Boolean(PrimitiveType):
+
+    def to_string(self):
+        return "bool"
+
+
+class NumberType(PrimitiveType):
+    pass
+
+
+class IntegerType(NumberType):
+    pass
+
+
+class SignedIntegerType(IntegerType):
+    pass
+
+
+class Int8(SignedIntegerType):
+
+    def to_string(self):
+        return "int8"
+
+
+class Int16(SignedIntegerType):
+
+    def to_string(self):
+        return "int16"
+
+
+class Int32(SignedIntegerType):
+
+    def to_string(self):
+        return "int32"
+
+
+class Int64(SignedIntegerType):
+
+    def to_string(self):
+        return "int64"
+
+
+class Binary(PrimitiveType):
+    """
+    A variable-size binary (bytes) value
+    """
+    def to_string(self):
+        return "binary"
+
+
+class String(PrimitiveType):
+    """
+    A UTF8-encoded string value
+    """
+    def to_string(self):
+        return "string"
+
+
+class Object(PrimitiveType):
+    """
+    Any PyObject value
+    """
+    def to_string(self):
+        return "object"
+
+
+class Categorical(DataType):
+    """
+    A categorical value is an ordinal (integer) value that references a
+    sequence of category values of an arbitrary data type
+    """
+
+    def __init__(self, index_type: IntegerType, category_type: DataType,
+                 ordered: bool = False):
+        self.index_type = index_type
+        self.category_type = category_type
+        self.ordered = ordered
+
+    def equals(self, other: DataType) -> bool:
+        return (isinstance(other, Categorical) and
+                self.index_type == other.index_type and
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    @property
+    def attrs(self) -> Mapping:
+        """
+        Metadata for this column. Default implementation returns empty dict
+        """
+        return {}
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
```

Comment:
Could these APIs be generalized more to something like `to_array` and `to_masked_array`? I imagine users will use these APIs as building blocks and if we're for example using a GPU we wouldn't want to either return a different typed GPU-backed object or cause an unnecessary host --> device copy.

For example, Pandas uses `.values` to return a numpy / Pandas array whereas in cuDF we return a cupy array to keep everything on the GPU as much as possible.

Reply:
I don't understand your comment. 

The idea of this interface is that the data representation is not known. So we will add various methods to `Column` (what is here is NOT exhaustive, maybe that is where there is confusion) to allow the consumer to request the data be returned to them in the representation that they prefer.

Reply:
I guess my point is if we have `to_numpy` and `to_cupy`, the standard will become `to_numpy` where anyone then looking to change to a different array implementation will have to go back and rework their code. If it was instead `to_array` and the object returned is expected to return something supporting some form of array protocols, it allows for writing more general code that can be run anywhere.

Maybe I was unclear as to whether `to_numpy` and `to_arrow` were placeholders for generic conversion functions for somewhat generic objects or if they were specific to returning numpy arrays and arrow arrays.

Reply:
I see. There is no "standard" here, and this effort is not trying to establish a standard representation. The idea is that some projects are implemented against a certain data representation (e.g. NumPy arrays) -- so indeed these methods are here so that a producer can return their data to the consumer in the representation that they are looking. matplotlib is an example -- it deals almost exclusively in NumPy arrays so rather than have matplotlib take responsibility for the details of converting from some foreign data frame object to a NumPy array, this conversion is the responsibility of the producer of this interface.

Reply:
> I guess my point is if we have to_numpy and to_cupy, the standard will become to_numpy where anyone then looking to change to a different array implementation will have to go back and rework their code.

Also for clarity, efforts to achieve duck-typing between array libraries is pretty orthogonal to this project -- if such a duck-typed array standard is developed and adopted / normalized, then this interface can add it. The use case here is strictly abstract data interchange that does not force producers to convert to a particular intermediate format like pandas.

Reply:
I'm not proposing for ducktyping, I'm proposing for adopting protocols instead of functions. I.E. instead of `to_numpy` we could expose `__array_interface__` and/or buffer protocol. Instead of `to_arrow` we could expose `__arrow_array__` . Instead of `to_cupy` we could expose `__cuda_array_interface__`.

With these inplace instead of calling `col.to_numpy()` it would be `numpy.asarray(col)` or instead of `col.to_arrow()` it would be `pyarrow.asarray(col)`.

Again this is based on your above examples of numpy and arrow, but is there situations where there's not a protocol that we could use to accomplish this?

Reply:
This object can certainly have the `__*_interface__` methods. That use cases can be fully served by these methods may be oversimplifying, though. Consider pandas's `to_numpy` method for example:

https://github.com/pandas-dev/pandas/blob/7763390a464bfff61b7b655f3acbc5fd70a0f08d/pandas/core/arrays/numpy_.py#L439

In many cases the conversion is going to be lossy, and so accessor methods will likely require some arguments/options.

Reply:
Understood, that makes sense. My fear is if protocols can't be relied on entirely, the default path consumers will take is to use `.to_numpy()` where ideally they could use something more generic like `.to_array()` which produces something array-like which supports calling numpy functions against it. Without something like this, a GPU library like cuDF that will want to support this protocol would have to make an awful choice of either returning something like a cupy array from the `to_numpy()` call to keep data on the GPU and do things efficiently but possibly break for some people / libraries integrating downstream who were expecting specifically host arrays or have `to_numpy()` explicitly copy the data to host, which then possibly throws away the performance we may have gained from using the GPU.

Reply:
I see. I don't know that it makes sense to be opinionated in this way at this level. We can provide people with choices and they can decide what makes sense for them. I don't see a problem with having a `to_$something` method as long as that `$something` is documented and something that people can understand how to use. So you can provide both the "general array like" path and the "NumPy" path and people can choose. That you have a preference about which one you want people to use seems like it shouldn't factor into the interface

Reply:
The problem is if the "NumPy" path is included in the interface whereas the "general array like" path isn't, most producers will likely only implement the "NumPy" path which will make almost all consumers consume the "NumPy" path. This then becomes a problem for us trying to integrate efficiently of trying to convince everyone to change to the "general array like" path as opposed to that being the typical path.

Reply:
Would you like to propose an API to add? In the PR description I wrote "The APIs proposed here are not intended to be to the exclusion of others -- more APIs can be proposed and added later." What is written here is not exhaustive.

Reply:
Apologies, I missed that. Will add a suggestion to kickoff discussions on adding the generic API I was talking about.